src/main/java/org/truffleruby/parser/TranslatorEnvironment.java

```diff
@@ -234,4 +234,8 @@ public void setBreakIDForWhile(BreakID breakID) {
     public LexicalScope unsafeGetLexicalScope() {
         return parseEnvironment.getLexicalScope();
     }
+
+    public boolean shouldWarnYieldInSingletonClass() {
+        return isModuleBody && !isBlock();
```

Comment:
I thought a more elaborate check would be needed, but actually it's indeed warned for any `yield` in a module body, not just `class << ...`:
```
$ ruby -we 'class << self; yield; end'
-e:1: warning: `yield' in class syntax will not be supported from Ruby 3.0. [Feature #15575]
Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `singleton class': no block given (yield) (LocalJumpError)

$ ruby -we 'module A; yield; end'
-e:1: warning: `yield' in class syntax will not be supported from Ruby 3.0. [Feature #15575]
Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `<module:A>': no block given (yield) (LocalJumpError)
```

Reply:
`isModuleBody` is always false when `isBlock()` is true.

The helper I was mentioning is something like:
```java
public TranslatorEnvironment getSurroundingMethodEnvironment() {
 TranslatorEnvironment methodParent = this; 
 while (methodParent.isBlock()) { 
     methodParent = methodParent.getParent(); 
 } 
 return methodParent;
}
```
(which should also be used in https://github.com/oracle/truffleruby/blob/d9235b8b2fd7294b7aa215e1bb05f55f31bee60a/src/main/java/org/truffleruby/parser/BodyTranslator.java#L1953 to avoid code duplication)

and then `shouldWarnYieldInSingletonClass` should be renamed to `shouldWarnYieldInModuleBody` and be like:
```java
    public boolean shouldWarnYieldInModuleBody() {
        return getSurroundingMethodEnvironment().isModuleBody();
    }
```

so that such a case is also warned:
```
$ ruby -we 'module A; 1.times { yield }; end'
-e:1: warning: `yield' in class syntax will not be supported from Ruby 3.0. [Feature #15575]
...
```